src/name_bazaar/server/emailer/listeners.cljs

```diff
@@ -0,0 +1,119 @@
+      (when-let [to-email (validate-email owner-encrypted-email)]
+        (sendgrid/send-notification-email {:from-email "hello@district0x.io"
+                                           :to-email to-email
+                                           :subject "Your offering was bought"
+                                           :content (templates/on-offering-bought offering name price)}
+                                          #(prn "Success sending on-offering-bought email")
+                                          #(prn "An error has occured: " %))))))
+
+(defn on-offering-changed [server-state {:keys [:offering :version :event-type :extra-data] :as args}]
+  (go
+    (let [{:keys [:name :original-owner :price :end-time :winning-bidder] :as result} (<! (offering-api/get-offering @server-state offering))]
+      (if winning-bidder
+        (on-auction-finalized server-state offering original-owner winning-bidder name price)
+        (on-offering-bought server-state offering original-owner name price)))))
+
+(defn on-new-bid
+  [server-state {:keys [:offering] :as args}]
+  (go
+    (let [{:keys [:name :original-owner :price :end-time :winning-bidder] :as result} (<! (offering-api/get-offering @server-state offering))
+          [_ owner-encrypted-email] (<! (district0x-emails-api/get-email @server-state {:district0x-emails/address original-owner}))]
+      (when-let [to-email (validate-email owner-encrypted-email)]
+        (sendgrid/send-notification-email {:from-email "hello@district0x.io"
+                                           :to-email to-email
+                                           :subject "Your offering received a new bid"
+                                           :content (templates/on-new-bid offering name price)}
+                                          #(prn "Success sending on-new-bid email")
+                                          #(prn "An error has occured: " %))))))
+
+(defn setup-listener!
+  ([server-state contract-key event-key callback]
+   (setup-listener! server-state contract-key event-key true nil callback))
+  ([server-state contract-key event-key retrieve-events? event-type callback]
+   (web3-eth/contract-call (state/instance @server-state contract-key)
+                           event-key
+                           (if event-type
+                             {:event-type event-type}
+                             {})
+                           (if retrieve-events?
+                             {:from-block 0 :to-block "latest"}
+                             "latest")
+                           (fn [error {:keys [:args] :as response}]
+                             (if error
+                               (prn "An error has occured: " error)
+                               (callback server-state args))))))
+
+(defn setup-event-listeners! [server-state]
+  (setup-listener! server-state :offering-registry :on-offering-added on-offering-added)
```

Comment:
You actually should keep track of running listeners, aka `filters` and every time `setup-event-listeners!` is run it should stop existing listeners if existent. You're launching new listeners every time figwheel reloads code and old ones keep running as well. Please take a look at `db_sync.cljs` on how this is solved.